### ISSUE TITLE
Burn inference results for the PaddleOCR v4 det model are inconsistent with ONNXRuntime.

### ISSUE THREAD

**Issue by yuanyan3060** [tracel-ai/burn-onnx#268]
**Describe the bug**
Burn inference results for the PaddleOCR v4 det model are inconsistent with ONNXRuntime.
You can find the example at https://github.com/yuanyan3060/burn-test.

input image
<img width="640" height="640" alt="Image" src="https://github.com/user-attachments/assets/04b6efc8-b52b-4805-8320-498c8ee396f7" />

burn output
<img width="640" height="640" alt="Image" src="https://github.com/user-attachments/assets/5449a1ca-8b1a-445b-b372-fa44f00511a6" />

onnxruntime output
<img width="640" height="640" alt="Image" src="https://github.com/user-attachments/assets/c440bbc8-c4b0-439b-8908-204b628ac169" />

- OS: Windows
- Version 0.21.0-pre.2

**Comment by laggui** [comment]
Hmmm it could be a mismatch in node configs from burn-onnx, or numerical discrepancies for a specific operation. Hard to say without more details or investigation, but the former is more likely.

It would be helpful to check numerical results without post-processing for better comparison.

**Comment by antimora** [comment]
Transferred to burn-onnx repo

**Comment by antimora** [comment]
## Investigation Summary

I analyzed the PaddleOCR v4 det model (504 nodes, opset 16) and verified each operator used against the burn-onnx codegen.

### Operators Verified Correct

All operator patterns used by this model produce correct numerical results when tested end-to-end:

- **Conv2d** (including depthwise with `groups=channels`): weight layout `[out, in/groups, kH, kW]` matches between ONNX and Burn, padding order correct
- **HardSwish / HardSigmoid**: formulas match ONNX spec, alpha/beta values passed correctly (model uses both `alpha=1/6` and `alpha=0.2`)
- **Resize** (nearest, asymmetric, floor): existing integration test uses identical config and passes
- **ConvTranspose2d**: weight layout matches, `pads=[0,0,0,0]` so no ordering issue
- **Scalar Mul/Add** (fused BN): shape `(1,)` constants correctly broadcast via `unsqueeze_dims([0,1,2])` to `[1,1,1,1]`
- **SE block**: residual pattern `x + x * HardSigmoid(SE(x))` correctly matches ONNX graph structure
- **BatchNorm, GlobalAveragePool, Sigmoid, Concat, Relu**: standard implementations

I also added a new integration test (`paddle_pattern`) that mimics the key PaddleOCR patterns (depthwise conv, scalar fused BN, HardSwish, SE block with HardSigmoid, ConvTranspose, Sigmoid) with all weights as Constant nodes (no initializers). It passes with `rel_err < 1e-4` vs the ONNX reference evaluator.

### Likely Root Cause: Weight Loading

Looking at the reproducer code in https://github.com/yuanyan3060/burn-test:

```rust
let model = det::Model::<Wgpu>::new(&device);
```

`Model::new(&device)` creates the model **without loading pretrained weights**. To load the `.bpk` weights, use one of:

```rust
// Option 1: Default (uses hardcoded path from build)
let model = det::Model::<Wgpu>::default();

// Option 2: Explicit path
let model = det::Model::<Wgpu>::from_file("path/to/model.bpk", &device);
```

With `new()`, Conv2d gets random Kaiming initialization and the Constant parameters (scalar BN scale/bias) initialize to zeros, which would collapse intermediate activations.

### Additional Note

The reproducer also has a version mismatch: `burn-onnx = "0.21.0-pre.1"` (build dep) vs `burn = "0.21.0-pre.2"` (runtime). These should match to avoid API incompatibilities.

### If the issue persists after fixing weight loading

1. Try the NdArray backend instead of WGPU to isolate whether it's a backend-specific numerical precision issue
2. Compare intermediate tensor values at key points (after Resize, after ConvTranspose) between Burn and ONNXRuntime

**Comment by antimora** [comment]
## Confirmed: Reproduced and Fixed

I cloned the [reproducer repo](https://github.com/yuanyan3060/burn-test) and ran it with the fix. The root cause is confirmed: **`Model::new(&device)` does not load pretrained weights**.

### Fix

Change line 88 in `src/main.rs`:

```diff
- let model = det::Model::<Wgpu>::new(&device);
+ let model = det::Model::<Wgpu>::default();
```

`Model::new()` creates the model with random/zero-initialized parameters. `Model::default()` calls `from_file()` internally to load the `.bpk` weights.

### Numerical comparison (with proper weight loading)

| Backend | Raw output sum | Min | Max |
|---------|---------------|-----|-----|
| ONNX Runtime | 866.9901 | 0.000000 | 1.000000 |
| Burn NdArray | 867.0131 | 0.000000 | 1.000000 |
| Burn WGPU | 867.0132 | 0.000000 | 1.000000 |

All three produce visually identical output images with the text detection pattern matching.

### Additional note

The reproducer also has a version mismatch (`burn-onnx = "0.21.0-pre.1"` vs `burn = "0.21.0-pre.2"`). These should match. I tested with local burn-onnx (current main) against burn 0.21.0-pre.2.

---

That said, the API is confusing here. Having a public `new()` that creates a broken (uninitialized) model is a footgun. We should consider either renaming it to something like `new_uninitialized()`, making it `pub(crate)`, or documenting the distinction more prominently.

**Comment by antimora** [comment]
The issue is a user error (Model::new() vs Model::default()), not a burn-onnx bug.
